rkt: add the cp command

fryckbos opened the pull request:
rkt cp copies files from or into running pods.

Fixes #3531

fryckbos commented:
The build failed on

build-rkt-1.25.0+git/gopath/src/github.com/rkt/rkt/rkt/cp.go:26:2: cannot find package "github.com/coreos/rkt/common"

This does not happen when I use the coreos.com/rkt/builder:1.2.0 image to build rkt on my machine.

lucab commented:
> build-rkt-1.25.0+git/gopath/src/github.com/rkt/rkt/rkt/cp.go:26:2: cannot find package "github.com/coreos/rkt/common"

We recently moved the repo from `coreos/rkt` to `rkt/rkt`, it looks like you may just need to fix your imports.

euank commented:
ok to test

euank commented on Documentation/commands.md:
```diff
@@ -83,12 +84,13 @@ In addition to the flags used by individual `rkt` commands, `rkt` has a set of g
 - **http**: Allow HTTP connections. Be warned that this will send any credentials as clear text, allowing anybody with access to your network to obtain them. It will also perform no validation of the remote server, making it possible for an attacker to impersonate the remote server. This applies specifically to fetching images, signatures, and gpg pubkeys.
 - **image**: Disables verifying image signatures. If someone is able to replace the image on the server with a modified one or is in a position to impersonate the server, they will be able to force you to run arbitrary code.
 - **tls**: Accept any certificate from the server and any host name in that certificate. This will make it possible for attackers to spoof the remote server and provide malicious images.
+- **ondisk**: Disables verifying the integrity of the on-disk, rendered image before running. This significantly speeds up start time. If an attacker is able to modify the contents of your local filesystem, this will allow them to cause you to run arbitrary malicious code.
```
looks like a bad rebase or merge in there, this should be gone.

euank commented on Documentation/subcommands/cp.md:
```diff
@@ -0,0 +1,57 @@
+
+## Copying directories
+
+Copying a directory copies all subdirectories and files. If the directory or subdirectories that do not exist yet will be created inside the destination directory. Files that already exist will be overwritten.
```
should there be a flag to not clobber?

euank commented on rkt/cp.go:
```diff
@@ -0,0 +1,271 @@
+func runCp(cmd *cobra.Command, args []string) (exit int) {
+	if len(args) < 2 {
+		cmd.Usage()
+		return 255
```
nit, 254 for consistency.

Same for the other return values.

In this case it might make more sense to not use that for everything (since it can't be confused with application exit), but the error message should be enough to distinguish these error cases

fryckbos commented:
I fixed the merge, added the no-clobber and fixed the exit codes.

The test for KVM seems to work, I'm not sure whether that cover KVM completely.

About the namespaces, what would be the best way to test this ? Where could I find some code that shows how to break into the namespace ?

euank commented on tests/rkt_cp_test.go:
```diff
@@ -0,0 +1,74 @@
+		t.Fatalf("Waited for the prompt but not found : %v", err)
+	}
+
+	cpCmd := fmt.Sprintf(`/bin/sh -c "%s cp $(%s list | grep running | awk '{print $1}'):/inspect %s"`, ctx.Cmd(), ctx.Cmd(), tmpDir)
```
This shouldn't be necessary. There's a run variant (`runRktAndGetUUID`) that can be used instead of awk.

euank commented on tests/rkt_cp_test.go:
```diff
@@ -0,0 +1,74 @@
+	cpCmd := fmt.Sprintf(`/bin/sh -c "%s cp $(%s list | grep running | awk '{print $1}'):/inspect %s"`, ctx.Cmd(), ctx.Cmd(), tmpDir)
+	t.Logf("Executing command '%s'", cpCmd)
+
+	//if err := runRktAndCheckRegexOutput(t, cpCmd, ""); err != nil {
```
nit, should be deleted

euank commented on tests/rkt_cp_test.go:
```diff
@@ -0,0 +1,74 @@
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	if _, err := os.Stat(fmt.Sprintf("%s/inspect", tmpDir)); err != nil {
```
nit, `filepath.Join` instead of `Sprintf`

euank commented on tests/rkt_cp_test.go:
```diff
@@ -0,0 +1,74 @@
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	if err := child.SendLine("Bye"); err != nil {
```
I think it's actually okay to not tear down the container. The teardown path isn't under test and the `ctx.Cleanup()` call should take care of nuking the processes.

@lucab, double checking relying on cleanup is sane here.

euank commented on tests/rkt_cp_test.go:
```diff
@@ -0,0 +1,74 @@
+)
+
+// TestRktCp tests 'rkt cp'.
+// It copies a file from within a contain and checks whether the file is present.
```
nit: container

Should it also copy in the other direction?

euank commented on rkt/cp.go:
```diff
@@ -0,0 +1,281 @@
+		return err
+	}
+
+	err = os.Chmod(dst, si.Mode())
```
if we're going to chmod, we might as well chown too

euank commented on rkt/cp.go:
```diff
@@ -0,0 +1,281 @@
+}
+
+// copyTree recursively copies a directory tree, preserving permissions.
+// Symlinks are ignored and skipped.
```
This behaviour seems surprising.

If I do `rkt cp $uuid:/path/to/symlink /tmp/` then I wont' get an error, it'll exit success, but nothing will have been copied

euank commented on rkt/cp.go:
```diff
@@ -0,0 +1,281 @@
+		return err
+	}
+
+	if si.Mode()&os.ModeSymlink != 0 {
```
This won't work, will it? Since `os.Stat` follows symlinks, this won't ever be true. The `os.Stat` above would need to be an `Lstat` instead.

euank commented on rkt/cp.go:
```diff
@@ -0,0 +1,281 @@
+		}
+	}
+
+	in, err := os.Open(src)
```
There's racyness present in the current implementation which will lead to this frequently erroring on a system with lots of activity, and could result in this copying the wrong thing.

The `Stat` and such done above could refer to a different file than ultimately gets opened here, so the symlink check could be false, the file could have been deleted, etc.

To make this not racy, a single fd will have to be preserved throughout the validate/copy/permissions steps.

euank commented on Documentation/subcommands/cp.md:
```diff
@@ -0,0 +1,58 @@
+
+Copy a file or directory from or to a running pod, given the pod UUID.
+
+Files on the host can be referenced using absolute or relative paths. Files inside a pod are referenced using `UUID:ABSOLUTE_PATH`. Files can be copied from the host to a pod or the other way around. Copying files between pods is not supported. The destination path must reference an existing directory. File permissions are preserved on copy. Symlinks are not followed and skipped.
```
The 'destination must reference a directory' seems surprising for copying individual files (e.g. `rkt cp $uuid:/bin/bash /tmp/my-container-bash` seems like something people would expect to work).

The symlink behaviour also seems a little odd. There might be a better option than ignoring symlinks? Especially for copying a single file I think there can be a better option

euank commented:
Apologies for letting this fall off my radar for a couple weeks, @fryckbos 

So, the namespace bit:

The easiest way to show what doesn't work is something like this:
```sh
$ rkt run --uuid-file-save=/tmp/uuid --volume host-tmp,kind=host,source=/tmp docker://busybox --mount volume=host-tmp,target=/tmp -- -c 'while sleep 1; do echo "-----"; ls /tmp; done'

# The container prints that it sees the host `/tmp`. If you have files/dirs in /tmp, they'll be displayed but can't be copied out

$ echo 'hello world' > test
$ rkt cp ./test $(cat /tmp/uuid):/tmp/

# The container doesn't print the new '/tmp/test' file

$ ls /var/lib/rkt/pods/run/$(cat /tmp/uuid)/stage1/rootfs/opt/stage2/busybox/rootfs/tmp

# the right path, but in the host's mount namespace, does have that file. In the pod's mount namespace it's masked, so e.g. `rkt exec` and the container itself see a different view
```

Fixing this is actually quite hard!
On entering into a mount namespace, you have to pick a file present in that namespace to execute.. so if the container doesn't have any code that can do the copy, you're basically sunk.

The options are basically:
1. Mandate that the container has `tar` available for `cp` to work (this is what `kubectl cp` does btw)
2. Attempt to re-create the mount setup the container (e.g. in the host's or stage1's mount namespace), and then copy from that. This is what docker does I think!
3. Document that cp operations across mount namespaces won't work. This is obviously the easiest ;)


Hopefully that explanation made sense. I don't really like any of the options. A second opinion or opinions would be welcome, cc @lucab
